byzcoin/service.go

```diff
@@ -604,6 +604,42 @@ func (s *Service) GetSignerCounters(req *GetSignerCounters) (*GetSignerCountersR
 	return &resp, nil
 }
 
+// GetUpdates returns instances that have a newer versions than the ones
+// passed to it.
+func (s *Service) GetUpdates(pr *GetUpdatesRequest) (*GetUpdatesReply, error) {
+	sb := s.db().GetByID(pr.LatestBlockID)
+	if sb == nil {
+		return nil, xerrors.New("cannot find skipblock while getting proof")
+	}
+	if len(sb.ForwardLink) > 0 {
+		return nil, xerrors.New("can only give proofs for latest block")
```

Comment:
why don't you take a skipchainID instead of a skipblockID? Then the service can directly fetch the latest block.

Reply:
That would be an option then. The current implementation never sends the forwardlinks to prove the latest block is valid, as it supposes that the client already has the latest block. Currently the following happens:

- client calls the badly named `StreamingTransactions` and listens for new blocks
- every time a new block is received, the client sends a list of known instances to the node, with the request to send only the updates
- the node checks the versions, and only creates `InclusionProof`s for the instances that have a newer version

The main goal of `GetUpdates` is to use as few as possible of the network. So sending the block when not needed is an advantage.

A possible extension would be the flag `GUFSendProofForLatestBlock` which would fill out the already available `ForwardLinks` and the `Latest`.